polygon/p2p/fetcher.go

```diff
@@ -17,14 +16,49 @@ import (
 
 const responseTimeout = 5 * time.Second
 
-var invalidFetchHeadersRangeErr = errors.New("invalid fetch headers range")
+var (
+	ErrInvalidFetchHeadersRange          = errors.New("invalid fetch headers range")
+	ErrIncompleteDownloadHeadersResponse = errors.New("incomplete download headers response")
+)
 
 type RequestIdGenerator func() uint64
 
 type Fetcher interface {
 	FetchHeaders(ctx context.Context, start uint64, end uint64, peerId PeerId) ([]*types.Header, error)
 }
 
+func NewTrackingFetcher(
+	logger log.Logger,
+	messageListener MessageListener,
+	messageSender MessageSender,
+	peerPenalizer PeerPenalizer,
+	requestIdGenerator RequestIdGenerator,
+	peerTracker PeerTracker,
+) Fetcher {
+	return &trackingFetcher{
+		Fetcher:     NewFetcher(logger, messageListener, messageSender, peerPenalizer, requestIdGenerator),
+		peerTracker: peerTracker,
+	}
+}
+
+type trackingFetcher struct {
+	Fetcher
+	peerTracker PeerTracker
+}
+
+func (tf *trackingFetcher) FetchHeaders(ctx context.Context, start uint64, end uint64, peerId PeerId) ([]*types.Header, error) {
+	res, err := tf.Fetcher.FetchHeaders(ctx, start, end, peerId)
+	if err != nil {
+		if errors.Is(err, context.DeadlineExceeded) || errors.Is(err, ErrIncompleteDownloadHeadersResponse) {
+			tf.peerTracker.BlockNumMissing(peerId, start)
```

Comment:
does it return some headers in case of ErrIncompleteDownloadHeadersResponse?
if it does, should track the last returned block + 1

Reply:
yeah, I was thinking of changing that error to a struct err so that we can get back that info - working on it